app/src/main/java/app/revanced/integrations/shared/GmsCoreSupport.java

```diff
@@ -48,40 +49,78 @@ private static void open(String queryOrLink, String message) {
-            Logger.printInfo(() -> "GmsCore was not found", exception);
-            open(getGmsCoreDownload(), str("gms_core_not_installed_warning"));
-            return;
+            Logger.printDebug(() -> "GmsCore was not found");
+            Utils.showToastLong(str("gms_core_not_installed_warning"));
+            open(getGmsCoreDownload());
+        } catch (Exception ex) {
+            Logger.printException(() -> "checkAvailability failure", ex);
         }
+    }
 
-        try (var client = context.getContentResolver().acquireContentProviderClient(GMS_CORE_PROVIDER)) {
-            if (client != null) return;
+    private static void showDoNotKillMyAppDialog(Context context, String messageKey) {
+        // Use a delay to allow the activity to finish initializing.
+        // Otherwise, if device is in dark mode the dialog is missing a dark mode color scheme.
+        Utils.runOnMainThreadDelayed(() -> {
+            new AlertDialog.Builder(context)
+                    .setIcon(android.R.drawable.ic_dialog_alert)
+                    .setTitle(str("gms_core_not_whitelisted_title"))
+                    .setMessage(str(messageKey))
+                    .setPositiveButton(android.R.string.ok, (dialog, id) -> {
+                        open(DONT_KILL_MY_APP_LINK);
+                        System.exit(0);
+                    })
+                    .setCancelable(false)
+                    // Do not use .setCancelable(), so if something is wrong
+                    // the user can use back button to dismiss the dialog (without shutting down).
+                    .show();
+        }, 100);
+    }
+
+    /**
+     * Injection point.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.N)
+    public static void checkGmsCoreWhitelisted(Activity context) {
+        try {
+            // Check, if GmsCore is whitelisted from battery optimizations.
+            var powerManager = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
+            if (!powerManager.isIgnoringBatteryOptimizations(GMS_CORE_PACKAGE_NAME)) {
+                Logger.printDebug(() -> "GmsCore is not whitelisted from battery optimizations");
+                showDoNotKillMyAppDialog(context, "gms_core_not_whitelisted_using_battery_optimizations_message");
+                return;
+            }
 
-            Logger.printInfo(() -> "GmsCore is not running in the background");
-            open(DONT_KILL_MY_APP_LINK, str("gms_core_not_running_warning"));
+            // Check, if GmsCore is running in the background.
+            try (var client = context.getContentResolver().acquireContentProviderClient(GMS_CORE_PROVIDER)) {
+                if (client == null) {
+                    Logger.printDebug(() -> "GmsCore is not running in the background");
+                    showDoNotKillMyAppDialog(context, "gms_core_not_whitelisted_not_allowed_in_background_message");
+                }
+            }
```

Comment:
I kept this check because I am assuming that regardless of fulfilling the check above regarding battery optimizations, additional changes may be necessary. If this is not the case, this check can be removed because the check above would make this one obsolete.